Sources/Basic/OutputByteStream.swift

```diff
@@ -55,32 +58,108 @@ public class OutputByteStream: OutputStream {
         return buffer.count
     }
 
-    /// The contents of the output stream.
-    ///
-    /// This method implicitly flushes the stream.
-    public var bytes: ByteString {
-        flush()
-        return ByteString(self.buffer)
+    /// Currently available buffer size.
+    private var availableBufferSize: Int {
+        return max(buffer.capacity, OutputByteStream.bufferSize) - buffer.count
     }
-    
+
+    /// Clears the buffer maintaining current capacity.
+    private func clearBuffer() {
+        buffer.removeAll(keepingCapacity: true)
+    }
+
     // MARK: Data Output API
 
     public func flush() {
-        // Do nothing.
+        writeImpl(ArraySlice(buffer))
+        buffer.removeAll(keepingCapacity: true)
     }
 
     /// Write an individual byte to the buffer.
     public func write(_ byte: UInt8) {
+        // If buffer is full, write and clear it.
+        if availableBufferSize == 0 {
+            writeImpl(ArraySlice(buffer))
+            clearBuffer()
+        }
+
         buffer.append(byte)
     }
 
     /// Write the contents of a UnsafeBufferPointer<UInt8>.
     func write(_ ptr: UnsafeBufferPointer<UInt8>) {
+        // FIXME: This entire implementation is almost a copy of `write(_ bytes: ArraySlice<UInt8>)`.
+        // Find a way to abstract.
+        let availableBufferSize = self.availableBufferSize
+
+        if ptr.count > availableBufferSize {
+            if buffer.isEmpty {
+                let bytesToWrite = ptr.count - (ptr.count % availableBufferSize)
+                // FIXME: This will probably copy the bytes.
+                writeImpl(ArraySlice(ptr[0..<bytesToWrite]))
```

Comment:
@ddunbar one option here is adding another `writeImpl` which takes `UnsafeBufferPointer<UInt8>`. Which do you prefer?